clippy_lints/src/matches/mod.rs

```diff
@@ -1118,8 +1119,13 @@ impl<'tcx> LateLintPass<'tcx> for Matches {
                 );
                 needless_match::check_if_let(cx, expr, &if_let);
             }
-        } else if !from_expansion {
-            redundant_pattern_match::check(cx, expr);
+        } else {
+            if let Some(while_let) = higher::WhileLet::hir(expr) {
+                significant_drop_in_scrutinee::check_while_let(cx, expr, while_let.let_expr, while_let.if_then);
+            }
+            if !from_expansion {
+                redundant_pattern_match::check(cx, expr);
+            }
         }
```

Comment:
nit (feel free to ignore), could this be written as:

```rust
} else if let Some(while_let) = higher::WhileLet::hir(expr) {
    significant_drop_in_scrutinee::check_while_let(cx, expr, while_let.let_expr, while_let.if_then);

    if !from_expansion {
        redundant_pattern_match::check(cx, while_let);
    }
}
```

and remove that redundant `higher::WhileLet::hir(expr)` in `redundant_pattern_match::check`, to improve readability for a little bit, plus it's a "redundant pattern matching" 😂. Unless that `check` function is called somewhere else.

Reply:
Thanks for your comments. Well, your suggestion makes sense to me.

However, after having a closer look at `redundant_pattern_match::check`, I find out that this method needs `expr` to be passed to `find_method_sugg_for_if_let` (which in turn needs at least `expr.span` and `expr.hir_id`). It seems that such information is not available in `WhileLet`. I'm afraid it might not be worth refactoring such methods?

https://github.com/rust-lang/rust-clippy/blob/48686adf48e2017feb9da59fbb606eb3c57dfe76/clippy_lints/src/matches/redundant_pattern_match.rs#L27

https://github.com/rust-lang/rust-clippy/blob/48686adf48e2017feb9da59fbb606eb3c57dfe76/clippy_lints/src/matches/redundant_pattern_match.rs#L173-L180